packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  // If there is no ticket creation date or no actions, we return an empty timeline.
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const sortedActions: rawUserAction[] = sortActions(actions);
+
+  const mappedUserActions = mapUserActions(sortedActions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedUserActions,
+    ticketCreationDate
+  );
+
+  // Filter out custom actions to be excluded from the timeline
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+// Sorting actions by most recent
```

Comment:
Same remark, please keep the comments only for the cases where there an extremely strong reason behind it.

Reply:
This one I put it there so its obvious its sorted by most recent first since its kinda critical to the understanding of how we build the timeline. Let me know if you want to remove it still, I agree to remove comments, but this one I wonder if we dont want to keep it

Reply:
Then call your method `sortActionsByInverseChronologicalOrder` for example, and no need for the comment. Or `sortActionsByMostRecent` is also valid.

Reply:
btw this who whole file in my opinion will be much better with JS doc comments documenting the methods and the parameters passed to them.